docs/new-architecture-app-modules-android.md

```diff
@@ -0,0 +1,415 @@
+---
+id: new-architecture-app-modules-android
+title: Enabling TurboMudule on Android
+---
+
+Make sure your application meets all the [prerequisites](new-architecture-app-intro).
+
+## 1. Enable NDK and the native build
+
+:::caution
+
+In this iteration of the playbook we’re setting up the project to let you build from source. You might notice an increase in your build time because of this. We’re looking into what would be the preferred approach here so please feel free to share your feedbacks.
+
+:::
+
+The code-gen will output some Java and some C++ code that now we need to build.
+
+Let’s edit your **module level** `build.gradle` to include the **two** `externalNativeBuild` blocks detailed below inside the `android{}` block:
+
+```groovy
+android {
+    defaultConfig {
+        applicationId "com.awesomeproject"
+        // ...
+
+        // Add this block
+        externalNativeBuild {
+            ndkBuild {
+                arguments "APP_PLATFORM=android-21",
+                        "APP_STL=c++_shared",
+                        "NDK_TOOLCHAIN_VERSION=clang",
+                        "GENERATED_SRC_DIR=$buildDir/generated/source",
+                        "PROJECT_BUILD_DIR=$buildDir",
+                        "REACT_ANDROID_DIR=$rootDir/../node_modules/react-native/ReactAndroid",
+                        "REACT_ANDROID_BUILD_DIR=$rootDir/../node_modules/react-native/ReactAndroid/build"
+                cFlags "-Wall", "-Werror", "-fexceptions", "-frtti", "-DWITH_INSPECTOR=1"
+                cppFlags "-std=c++17"
+                targets "myapplication_appmodules"
+            }
+        }
+    }
+
+    // Add this block
+    externalNativeBuild {
+        ndkBuild {
+            path "$projectDir/src/main/jni/Android.mk"
+        }
+    }
+}
+```
+
+In the same `build.gradle` file, inside the same `android{}` let’s add also the following section:
+
+```groovy
+android {
+    // ...
+
+    def reactAndroidProjectDir = project(':ReactAndroid').projectDir
+    def packageReactNdkLibs = tasks.register("packageReactNdkLibs", Copy) {
+        dependsOn(":ReactAndroid:packageReactNdkLibsForBuck")
+        dependsOn("generateCodegenArtifactsFromSchema")
+        from("$reactAndroidProjectDir/src/main/jni/prebuilt/lib")
+        into("$buildDir/react-ndk/exported")
+    }
+
+    afterEvaluate {
+        preBuild.dependsOn(packageReactNdkLibs)
+    }
+
+    packagingOptions {
+        pickFirst '**/libhermes.so'
+        pickFirst '**/libjsc.so'
+    }
+}
+```
+
+Finally, we need to create a Makefile inside the `src/main/jni` folder called `Android.mk` with the following content:
```

Comment:
I am using react-native@0.0.0-20220120-2008-2f67f5d68

https://github.com/nomi9995/react-native-turbomodule-fabric-renderer-boilerpate

I am always getting this error from `android.mk`, is there anything, I am doing wrong?
```
C/C++: fcntl(): Bad file descriptor
C/C++: /Users/apple/Library/Android/sdk/ndk/21.4.7075529/build/core/prebuilt-library.mk:45: *** Android NDK: Aborting    .  Stop.
[CXX1405] error when building with ndkBuild using /Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/src/main/jni/Android.mk: Build command failed.
Error while executing process /Users/apple/Library/Android/sdk/ndk/21.4.7075529/ndk-build with arguments {NDK_PROJECT_PATH=null APP_BUILD_SCRIPT=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/src/main/jni/Android.mk APP_ABI=x86 NDK_ALL_ABIS=x86 NDK_DEBUG=1 NDK_OUT=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/build/intermediates/cxx/Debug/1e4i2923/obj NDK_LIBS_OUT=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/build/intermediates/cxx/Debug/1e4i2923/lib APP_CFLAGS+=-Wall APP_CFLAGS+=-Werror APP_CFLAGS+=-fexceptions APP_CFLAGS+=-frtti APP_CFLAGS+=-DWITH_INSPECTOR=1 APP_CPPFLAGS+=-std=c++17 APP_PLATFORM=android-21 APP_STL=c++_shared NDK_TOOLCHAIN_VERSION=clang GENERATED_SRC_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/build/generated/source PROJECT_BUILD_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/build REACT_ANDROID_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/../node_modules/react-native/ReactAndroid REACT_ANDROID_BUILD_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/../node_modules/react-native/ReactAndroid/build APP_SHORT_COMMANDS=false LOCAL_SHORT_COMMANDS=false -B -n}
Android NDK: ERROR:/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/src/main/jni/Android.mk:fb: LOCAL_SRC_FILES points to a missing file    
Android NDK: Check that /Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/build/react-ndk/exported/x86/libfb.so exists  or that its path is correct   

fcntl(): Bad file descriptor
/Users/apple/Library/Android/sdk/ndk/21.4.7075529/build/core/prebuilt-library.mk:45: *** Android NDK: Aborting    .  Stop.

C/C++: /Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/../ReactCommon/butter/Android.mk:28: *** Android NDK: Aborting.    .  Stop.
[CXX1405] error when building with ndkBuild using /Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/src/main/jni/react/jni/Android.mk: Build command failed.
Error while executing process /Users/apple/Library/Android/sdk/ndk/21.4.7075529/ndk-build with arguments {NDK_PROJECT_PATH=null APP_BUILD_SCRIPT=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/src/main/jni/react/jni/Android.mk APP_ABI=x86 NDK_ALL_ABIS=x86 NDK_DEBUG=1 APP_PLATFORM=android-21 NDK_OUT=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/build/intermediates/cxx/Debug/681pv6ab/obj NDK_LIBS_OUT=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/build/intermediates/cxx/Debug/681pv6ab/lib NDK_APPLICATION_MK=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/src/main/jni/Application.mk THIRD_PARTY_NDK_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/build/third-party-ndk REACT_COMMON_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/../ReactCommon REACT_GENERATED_SRC_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/build/generated/source REACT_SRC_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react -j12 --output-sync=none APP_SHORT_COMMANDS=false LOCAL_SHORT_COMMANDS=false -B -n}
Android NDK: WARNING: Ignoring unknown import directory: /Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/build/generated/source    
Android NDK: /Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/../ReactCommon/butter/Android.mk: Cannot find module with tag 'glog' in import path    
Android NDK: Are you sure your NDK_MODULE_PATH variable is properly defined ?    
Android NDK: The following directories were searched:    
Android NDK:         

/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/../ReactCommon/butter/Android.mk:28: *** Android NDK: Aborting.    .  Stop.
```

Reply:
> I am always getting this error from `android.mk`, is there anything, I am doing wrong?

Thanks for reporting this. This is definitely a bug on Google's end for AGP: https://issuetracker.google.com/issues/207403732

For now, you should be good to go with the following in your `android/app/build.gradle`:

```diff
afterEvaluate {
    preBuild.dependsOn(packageReactNdkLibs)
+   configureNdkBuildDebug.dependsOn(preBuild)
+   configureNdkBuildRelease.dependsOn(preBuild)
}
```

I've tested it on your project with `yarn && cd android && ./gradlew assembleDebug` and it works on my end.

I will also update the playbook to reflect this.

Reply:
@cortinico after resolving the above issue, I am getting another issue
- `codegen` is not creating `com.example.samplelibrary.MyApplicationLibrarySpec` on gradle sync

![image](https://user-images.githubusercontent.com/36044436/150548680-13f88d98-f351-4475-b6cd-09453b918729.png)

- if i delete `android/app/build/react-ndk` than `react-ndk` is not auto creating if i do gradle sync, rebuild project etc

Reply:
> on gradle sync

Nothing of the codegen gets built with a Gradle Sync. We haven't (yet) looked into how to plug this step inside the IDE/Gradle Sync phase. If you invoke `assembleDebug` from the command line OR invoke `yarn react-native run-android` everything should work.

Reply:
@cortinico one again, thanks, if codegen on Gradle sync is required then I can contribute this one because I have already done this type of task in my open source project 
https://github.com/nomi9995/react-native-keys/blob/master/android/RNKeys.gradle

can you confirm for me if this codegen in Gradle sync is required in the future or not

Reply:
@cortinico another issue that I am getting when I try to run `./gradlew assemble`
```
/Users/apple/Data/Work/shobbak/reactnative/packages/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/ReactExtension.kt: (10, 37): 'BaseVariant' is deprecated. Deprecated in Java
```
![image](https://user-images.githubusercontent.com/36044436/150551975-e6d233f7-dd9c-49b7-8768-4f49dad1b0e0.png)

Reply:
> can you confirm for me if this codegen in Gradle sync is required in the future or not

We haven't discussed on this front yet (cc @ShikaSD). Thanks for your input though 👍 

> another issue that I am getting when I try to run ./gradlew assemble

That's a build warning that we had as we moved to AGP 7.x as the `'BaseVariant'` is deprecated and we should move to use the newer variant API here. Not a blocker for this guide though.

Reply:
@cortinico I have updated react-native nightly to latest nightly release but i am unable to run the android app because - codegen is not creating files
- ios app is running but i am unable to call `getString` function in `App.tsx`
is there i am doing something wrong?

https://github.com/nomi9995/react-native-turbomodule-fabric-renderer-boilerpate

![image](https://user-images.githubusercontent.com/36044436/150652787-fb1f0971-0a06-4deb-8a8c-d7db415a0c0d.png)

Reply:
@cortinico ?